### PR TITLE
Add {VirtualHost, Server}.defaultHostname()

### PR DESCRIPTION
**Pull request by trustin**
- Add VirtualHost.defaultHostname
  - Unless specified, it is determined from the user-specified hostname
    pattern or the current machine's hostname
- Add Server.defaultHostname() which is a shortcut to
  config().defaultVirtualHost().defaultHostname()
- Use Server.defaultHostname() if a user does not specify hostname when
  creating JettyService or TomcatService

- Add missing configuration options to JettyServiceBuilder
- Merge TomcatServiceInvocationHandler and
  ManagedTomcatServiceInvocationHandler
  - Move the part related with the managed one to
    ManagedConnectorFactory
- Call Tomcat/Jetty Server.destroy() when it is managed and stopped.
- Fix UnsupportedOperationException in ThreadPoolExecutor.stop() when
  embedded Jetty is stopped
  - Add ArmeriaThreadPool
- Fix Tomcat LifeCycleExceptions while stopping Tomcats in
  TomcatServiceTest due to duplicate engine names
- Determine Tomcat engine name from service name when engine name was
  not specified

### CODE REVIEW

**Comment by trustin** [comment]
@blmarket Gentle ping

**Comment by blmarket** [comment]
hope `VirtualHostBuilder` have some unit tests.

One example:
```
new VirtualHostBuilder("*").build();
```
throws `IllegalArgumentException` as `defaultHostname` is `*`. expected was `localhost`(or equivalent).

File: src/main/java/com/linecorp/armeria/server/http/jetty/JettyServiceInvocationHandler.java [hunk 1]
```diff
@@ -54,17 +56,24 @@
...
+
+    private String hostname;
     private Server server;
     private ArmeriaConnector connector;
     private com.linecorp.armeria.server.Server armeriaServer;
     private boolean startedServer;
 
-    JettyServiceInvocationHandler(String hostname, Function<ExecutorService, Server> serverFactory) {
+    JettyServiceInvocationHandler(String hostname, Function<ExecutorService, Server> serverFactory,
+                                  boolean destroyOnStop) {
```
Comment (blmarket) [inline src/main/java/com/linecorp/armeria/server/http/jetty/JettyServiceInvocationHandler.java]: Not for this PR: If there is some callback for lifecycle events, we can replace this flag into decorator pattern, seems out of scope of this PR though. how do you think about this idea?

**Comment by blmarket** [comment]
left some nit comments, overall code looks great!

**Comment by trustin** [comment]
> hope `VirtualHostBuilder` have some unit tests.

Added `VirtualHostBuilderTest` and fixed one bug.

File: src/main/java/com/linecorp/armeria/server/http/jetty/JettyServiceInvocationHandler.java (hunk 1 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/http/jetty/JettyServiceInvocationHandler.java]: Sounds good to me. Let me just do it in this PR as well.

Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/http/jetty/JettyServiceInvocationHandler.java]: Replaced `destroyOnStop` with `postStopTask` (`Consumer`) for both Jetty and Tomcat.

**Comment by blmarket** [comment]
:+1:
